File: apis/backend-for-frontend/src/integrations/PlaylistsIntegration/translators.ts
import { Music, Playlist, PlaylistsList } from './proto/playlists_service_pb';
import MusicEntity from '@entities/Music';
import PlaylistEntity from '@entities/Playlist';

export function translatePlaylistEntity(playlist: Playlist): PlaylistEntity {
  const tracks = new Map<number, MusicEntity>();

  playlist.getTracksList().map(track => {
    const music = track.getMusic();
    const index = track.getIndex();

    if (music) {
      tracks.set(index, translateMusicEntity(music));
    }
  });

  return new PlaylistEntity({
    id: playlist.getId(),
    name: playlist.getName(),
    userId: playlist.getUserid(),
    tracks,
  });
}

export function translatePlaylistEntityList(playlistsList: PlaylistsList): Array<PlaylistEntity> {
  return playlistsList.getPlaylistsList().map(playlist => translatePlaylistEntity(playlist));
}

export function translateMusicEntity(music: Music): MusicEntity {
  return new MusicEntity({
    id: music.getId(),
    title: music.getTitle(),
    durationInSeconds: music.getDurationinseconds(),
    file: music.getFile(),
    composers: music.getComposersList(),
    lyrics: music.getLyrics(),
    albumId: music.getAlbumid(),
    views: music.getViews(),
  });
}

File: apis/playlists-service/internal/integrations/musicsIntegration/translators.go
package musicsIntegration

import (
	e "playlists-service/internal/entities"
	"playlists-service/internal/integrations/musicsIntegration/proto"
)

func translateMusicEntity(music *proto.Music) *e.Music {
	return &e.Music{
		Id:        music.Id,
		Title:     music.Title,
		Duration:  music.Duration,
		File:      music.File,
		Composers: music.Composers,
		Lyrics:    music.Lyrics,
		Views:     music.Views,
		AlbumId:   music.AlbumId,
		ArtistId:  music.ArtistId,
	}
}
